packages/ii-login-button/src/ii-login-button.ts

```diff
@@ -0,0 +1,277 @@
+ * @example
+ * button.addEventListener('ready', (e)=>{
+ *  // configure login and auth-client options here:
+ *  e.target.configure({
+ *    identityProvider: "http://localhost:3000"
+ *  });
+ * })
+ */
+export type IILoginButtonProps = {
+  createOptions?: AuthClientCreateOptions;
+  loginOptions?: AuthClientLoginOptions;
+  storage?: AuthClientStorage;
+};
+
+interface State {
+  authClient?: AuthClient;
+  isAuthenticated: boolean;
+  identity?: Identity;
+  ready: boolean;
+  createOptions?: AuthClientCreateOptions;
+  loginOptions?: AuthClientLoginOptions;
+  storage?: AuthClientStorage;
+}
+
+/**
+ * Internet Identity Login Button
+ * Drop-in web component to allow your users to log in with Internet Identity
+ *
+ * To use:
+ * Set up the component with <ii-login-button></ii-login-button>
+ * @example
+ * button.addEventListener('ready', (e)=>{
+ *  // configure login and auth-client options here:
+ *  e.target.configure(options);
+ * })
+ * button.addEventListener('login', (e)=>{
+ *  // Access authClient
+ *  const client = e.target.authClient;
+ *  // Principal
+ *  const principal = e.target.principal // also principalString
+ *  // AccountIdentifier for ICP ledger
+ *  const account = e.target.accountIdentifier
+ * })
+ */
+class IILoginButton extends LoginButton {
+  private _state: State;
+  constructor() {
+    super();
+    if (!this.shadowRoot) {
+      this.attachShadow({ mode: "open" });
+    }
+    this._state = {
+      isAuthenticated: false,
+      ready: false,
+    };
+
+    this.init();
+    this.render();
+  }
+
+  private init() {
+    AuthClient.create(this._state.createOptions)
+      .then(async (client) => {
+        this.setState({
+          ...this._state,
+          authClient: client,
+          isAuthenticated: await client.isAuthenticated(),
+          identity: await client.getIdentity(),
+        });
+        // fire ready event
+        if (!this._state.ready) {
+          const event = new ReadyEvent("ready");
+          this.dispatchEvent(event);
+          this.setState({ ready: true });
+        }
+      })
+      .catch((error) => {
+        console.error(error);
+        this.setState({ authClient: undefined });
+      });
+  }
+
+  connectedCallback() {
+    const login = this._login.bind(this);
+    this.shadowRoot
+      ?.querySelector("#ii-login-button")
+      ?.addEventListener("click", login);
+  }
+
+  setState(newState: { [Property in keyof State]+?: State[Property] }) {
+    this._state = {
+      ...this._state,
+      ...newState,
+    };
+  }
+
+  private render() {
+    const shadowRoot = this.shadowRoot as ShadowRoot;
+    const oldButton = shadowRoot.querySelector("button");
+    const newButton = document.createElement("button");
+    newButton.id = "ii-login-button";
+
+    const logoRight = this.hasAttribute("logo-right");
+    const label = this.getAttribute("label") ?? "Login With Internet Identity";
+    if (logoRight) {
```

Comment:
you can solve the alignment with CSS e.g. by a display `flex` reverse or not

Reply:
I could, but I vainly prefer shifting the actual semantic layout in the DOM

Reply:
If I may ask, is the semantic more meaningful if the logo finds place before or after the text, shouldn't it be set with a `role=presentation` in any case?

Generally speaking e.g. in NNS-dapp we avoid JS to shift element for stricly display purpose and solve this with CSS for performance and maintability reasons so just to curious to know if I am missing something.

Reply:
I'll accept that argument